File: src/test/resources/multilang/python/test_amqp.py
#!/usr/bin/env python3
"""
AMQP 0-9-1 Compliance Test using Python Pika library.
Tests basic AMQP operations against the server.
"""

import pika
import sys
import json
import time
import uuid
import traceback

class AmqpComplianceTest:
    def __init__(self, host='localhost', port=5672):
        self.host = host
        self.port = port
        self.results = []
        self.connection = None
        self.channel = None

    def connect(self):
        """Establish connection to AMQP server."""
        credentials = pika.PlainCredentials('guest', 'guest')
        parameters = pika.ConnectionParameters(
            host=self.host,
            port=self.port,
            credentials=credentials,
            connection_attempts=3,
            retry_delay=1,
            socket_timeout=10
        )
        self.connection = pika.BlockingConnection(parameters)
        self.channel = self.connection.channel()

    def disconnect(self):
        """Close connection."""
        if self.connection and self.connection.is_open:
            self.connection.close()

    def run_test(self, name, test_func):
        """Run a single test and record result."""
        try:
            test_func()
            self.results.append({'test': name, 'status': 'PASS', 'error': None})
            print(f"  [PASS] {name}")
            return True
        except Exception as e:
            self.results.append({'test': name, 'status': 'FAIL', 'error': str(e)})
            print(f"  [FAIL] {name}: {e}")
            traceback.print_exc()
            return False

    # ==================== Connection Tests ====================

    def test_connection_open(self):
        """Test that connection opens successfully."""
        assert self.connection.is_open, "Connection should be open"

    def test_channel_open(self):
        """Test that channel opens successfully."""
        assert self.channel.is_open, "Channel should be open"

    # ==================== Queue Tests ====================

    def test_queue_declare(self):
        """Test queue declaration."""
        queue_name = f"test.python.queue.{uuid.uuid4().hex[:8]}"
        result = self.channel.queue_declare(queue=queue_name, auto_delete=True)
        assert result.method.queue == queue_name

    def test_queue_declare_anonymous(self):
        """Test anonymous queue declaration."""
        result = self.channel.queue_declare(queue='', exclusive=True)
        assert result.method.queue != ''

    # ==================== Exchange Tests ====================

    def test_exchange_declare_direct(self):
        """Test direct exchange declaration."""
        exchange_name = f"test.python.direct.{uuid.uuid4().hex[:8]}"
        self.channel.exchange_declare(exchange=exchange_name, exchange_type='direct', auto_delete=True)

    def test_exchange_declare_fanout(self):
        """Test fanout exchange declaration."""
        exchange_name = f"test.python.fanout.{uuid.uuid4().hex[:8]}"
        self.channel.exchange_declare(exchange=exchange_name, exchange_type='fanout', auto_delete=True)

    def test_exchange_declare_topic(self):
        """Test topic exchange declaration."""
        exchange_name = f"test.python.topic.{uuid.uuid4().hex[:8]}"
        self.channel.exchange_declare(exchange=exchange_name, exchange_type='topic', auto_delete=True)

    # ==================== Binding Tests ====================

    def test_queue_bind(self):
        """Test queue binding to exchange."""
        exchange_name = f"test.python.bind.ex.{uuid.uuid4().hex[:8]}"
        queue_name = f"test.python.bind.q.{uuid.uuid4().hex[:8]}"

        self.channel.exchange_declare(exchange=exchange_name, exchange_type='direct', auto_delete=True)
        self.channel.queue_declare(queue=queue_name, auto_delete=True)
        self.channel.queue_bind(queue=queue_name, exchange=exchange_name, routing_key='test.key')

    # ==================== Publish/Consume Tests ====================

    def test_basic_publish_consume(self):
        """Test basic message publish and consume."""
        queue_name = f"test.python.pubsub.{uuid.uuid4().hex[:8]}"
        test_message = f"Hello from Python! {uuid.uuid4().hex}"
        received = []

        self.channel.queue_declare(queue=queue_name, auto_delete=True)
        self.channel.basic_publish(exchange='', routing_key=queue_name, body=test_message)

        def callback(ch, method, properties, body):
            received.append(body.decode())
            ch.basic_ack(delivery_tag=method.delivery_tag)

        self.channel.basic_consume(queue=queue_name, on_message_callback=callback, auto_ack=False)

        # Process one message with timeout
        start = time.time()
        while len(received) == 0 and time.time() - start < 5:
            self.connection.process_data_events(time_limit=0.5)

        assert len(received) == 1, f"Expected 1 message, got {len(received)}"
        assert received[0] == test_message, f"Message mismatch: {received[0]} != {test_message}"

    def test_publish_with_properties(self):
        """Test message publish with properties."""
        queue_name = f"test.python.props.{uuid.uuid4().hex[:8]}"
        received_props = [None]

        self.channel.queue_declare(queue=queue_name, auto_delete=True)

        properties = pika.BasicProperties(
            content_type='text/plain',
            correlation_id='corr-123',
            message_id='msg-456',
            priority=5
        )
        self.channel.basic_publish(exchange='', routing_key=queue_name, body='test', properties=properties)

        def callback(ch, method, props, body):
            received_props[0] = props
            ch.basic_ack(delivery_tag=method.delivery_tag)

        self.channel.basic_consume(queue=queue_name, on_message_callback=callback, auto_ack=False)

        start = time.time()
        while received_props[0] is None and time.time() - start < 5:
            self.connection.process_data_events(time_limit=0.5)

        assert received_props[0] is not None, "Should receive message properties"
        assert received_props[0].content_type == 'text/plain'
        assert received_props[0].correlation_id == 'corr-123'
        assert received_props[0].message_id == 'msg-456'

    def test_publish_with_headers(self):
        """Test message publish with headers."""
        queue_name = f"test.python.headers.{uuid.uuid4().hex[:8]}"
        received_headers = [None]

        self.channel.queue_declare(queue=queue_name, auto_delete=True)

        headers = {'x-custom-header': 'custom-value', 'x-number': 42}
        properties = pika.BasicProperties(headers=headers)
        self.channel.basic_publish(exchange='', routing_key=queue_name, body='test', properties=properties)

        def callback(ch, method, props, body):
            received_headers[0] = props.headers
            ch.basic_ack(delivery_tag=method.delivery_tag)

        self.channel.basic_consume(queue=queue_name, on_message_callback=callback, auto_ack=False)

        start = time.time()
        while received_headers[0] is None and time.time() - start < 5:
            self.connection.process_data_events(time_limit=0.5)

        assert received_headers[0] is not None, "Should receive headers"
        assert received_headers[0].get('x-custom-header') == 'custom-value'

    def test_message_ordering(self):
        """Test that messages are delivered in order."""
        queue_name = f"test.python.order.{uuid.uuid4().hex[:8]}"
        message_count = 10
        received = []

        self.channel.queue_declare(queue=queue_name, auto_delete=True)

        for i in range(message_count):
            self.channel.basic_publish(exchange='', routing_key=queue_name, body=f"Message-{i}")

        def callback(ch, method, props, body):
            received.append(body.decode())
            ch.basic_ack(delivery_tag=method.delivery_tag)

        self.channel.basic_consume(queue=queue_name, on_message_callback=callback, auto_ack=False)

        start = time.time()
        while len(received) < message_count and time.time() - start < 10:
            self.connection.process_data_events(time_limit=0.5)

        assert len(received) == message_count
        for i in range(message_count):
            assert received[i] == f"Message-{i}", f"Out of order at {i}"

    # ==================== Acknowledgment Tests ====================

    def test_basic_nack_requeue(self):
        """Test nack with requeue."""
        queue_name = f"test.python.nack.{uuid.uuid4().hex[:8]}"
        delivery_count = [0]

        self.channel.queue_declare(queue=queue_name, auto_delete=True)
        self.channel.basic_publish(exchange='', routing_key=queue_name, body='nack-me')

        def callback(ch, method, props, body):
            delivery_count[0] += 1
            if delivery_count[0] == 1:
                ch.basic_nack(delivery_tag=method.delivery_tag, requeue=True)
            else:
                ch.basic_ack(delivery_tag=method.delivery_tag)

        self.channel.basic_consume(queue=queue_name, on_message_callback=callback, auto_ack=False)

        start = time.time()
        while delivery_count[0] < 2 and time.time() - start < 5:
            self.connection.process_data_events(time_limit=0.5)

        assert delivery_count[0] == 2, f"Expected 2 deliveries, got {delivery_count[0]}"

    # ==================== QoS Tests ====================

    def test_basic_qos(self):
        """Test prefetch count."""
        queue_name = f"test.python.qos.{uuid.uuid4().hex[:8]}"

        self.channel.queue_declare(queue=queue_name, auto_delete=True)
        self.channel.basic_qos(prefetch_count=5)

        for i in range(10):
            self.channel.basic_publish(exchange='', routing_key=queue_name, body=f"Message-{i}")

    # ==================== Exchange Routing Tests ====================

    def test_direct_routing(self):
        """Test direct exchange routing."""
        exchange_name = f"test.python.routing.direct.{uuid.uuid4().hex[:8]}"
        queue1 = f"test.python.routing.q1.{uuid.uuid4().hex[:8]}"
        queue2 = f"test.python.routing.q2.{uuid.uuid4().hex[:8]}"
        received_q1 = []
        received_q2 = []

        self.channel.exchange_declare(exchange=exchange_name, exchange_type='direct', auto_delete=True)
        self.channel.queue_declare(queue=queue1, auto_delete=True)
        self.channel.queue_declare(queue=queue2, auto_delete=True)
        self.channel.queue_bind(queue=queue1, exchange=exchange_name, routing_key='key1')
        self.channel.queue_bind(queue=queue2, exchange=exchange_name, routing_key='key2')

        self.channel.basic_publish(exchange=exchange_name, routing_key='key1', body='to-q1')
        self.channel.basic_publish(exchange=exchange_name, routing_key='key2', body='to-q2')

        self.channel.basic_consume(queue=queue1, on_message_callback=lambda ch, m, p, b: received_q1.append(b.decode()) or ch.basic_ack(m.delivery_tag), auto_ack=False)
        self.channel.basic_consume(queue=queue2, on_message_callback=lambda ch, m, p, b: received_q2.append(b.decode()) or ch.basic_ack(m.delivery_tag), auto_ack=False)

        start = time.time()
        while (len(received_q1) < 1 or len(received_q2) < 1) and time.time() - start < 5:
            self.connection.process_data_events(time_limit=0.5)

        assert received_q1 == ['to-q1']
        assert received_q2 == ['to-q2']

    def test_fanout_routing(self):
        """Test fanout exchange routing."""
        exchange_name = f"test.python.routing.fanout.{uuid.uuid4().hex[:8]}"
        queue1 = f"test.python.fanout.q1.{uuid.uuid4().hex[:8]}"
        queue2 = f"test.python.fanout.q2.{uuid.uuid4().hex[:8]}"
        received_q1 = []
        received_q2 = []

        self.channel.exchange_declare(exchange=exchange_name, exchange_type='fanout', auto_delete=True)
        self.channel.queue_declare(queue=queue1, auto_delete=True)
        self.channel.queue_declare(queue=queue2, auto_delete=True)
        self.channel.queue_bind(queue=queue1, exchange=exchange_name, routing_key='')
        self.channel.queue_bind(queue=queue2, exchange=exchange_name, routing_key='')

        self.channel.basic_publish(exchange=exchange_name, routing_key='', body='broadcast')

        self.channel.basic_consume(queue=queue1, on_message_callback=lambda ch, m, p, b: received_q1.append(b.decode()) or ch.basic_ack(m.delivery_tag), auto_ack=False)
        self.channel.basic_consume(queue=queue2, on_message_callback=lambda ch, m, p, b: received_q2.append(b.decode()) or ch.basic_ack(m.delivery_tag), auto_ack=False)

        start = time.time()
        while (len(received_q1) < 1 or len(received_q2) < 1) and time.time() - start < 5:
            self.connection.process_data_events(time_limit=0.5)

        assert received_q1 == ['broadcast']
        assert received_q2 == ['broadcast']

    def run_all_tests(self):
        """Run all compliance tests."""
        print(f"\n{'='*60}")
        print("Python Pika AMQP 0-9-1 Compliance Tests")
        print(f"Server: {self.host}:{self.port}")
        print(f"{'='*60}\n")

        tests = [
            ('connection_open', self.test_connection_open),
            ('channel_open', self.test_channel_open),
            ('queue_declare', self.test_queue_declare),
            ('queue_declare_anonymous', self.test_queue_declare_anonymous),
            ('exchange_declare_direct', self.test_exchange_declare_direct),
            ('exchange_declare_fanout', self.test_exchange_declare_fanout),
            ('exchange_declare_topic', self.test_exchange_declare_topic),
            ('queue_bind', self.test_queue_bind),
            ('basic_publish_consume', self.test_basic_publish_consume),
            ('publish_with_properties', self.test_publish_with_properties),
            ('publish_with_headers', self.test_publish_with_headers),
            ('message_ordering', self.test_message_ordering),
            ('basic_nack_requeue', self.test_basic_nack_requeue),
            ('basic_qos', self.test_basic_qos),
            ('direct_routing', self.test_direct_routing),
            ('fanout_routing', self.test_fanout_routing),
        ]

        passed = 0
        failed = 0

        for name, test_func in tests:
            if self.run_test(name, test_func):
                passed += 1
            else:
                failed += 1

        print(f"\n{'='*60}")
        print(f"Results: {passed} passed, {failed} failed")
        print(f"{'='*60}\n")

        return {'passed': passed, 'failed': failed, 'results': self.results}


def main():
    host = sys.argv[1] if len(sys.argv) > 1 else 'localhost'
    port = int(sys.argv[2]) if len(sys.argv) > 2 else 5672

    test = AmqpComplianceTest(host, port)

    try:
        print(f"Connecting to AMQP server at {host}:{port}...")
        test.connect()
        results = test.run_all_tests()

        # Output JSON results for parsing
        print("\n--- JSON RESULTS ---")
        print(json.dumps(results))

        sys.exit(0 if results['failed'] == 0 else 1)
    except Exception as e:
        print(f"Connection failed: {e}")
        traceback.print_exc()
        sys.exit(1)
    finally:
        test.disconnect()


if __name__ == '__main__':
    main()

File: src/test/resources/multilang/nodejs/test_amqp.js
#!/usr/bin/env node
/**
 * AMQP 0-9-1 Compliance Test using Node.js amqplib library.
 * Tests basic AMQP operations against the server.
 */

const amqp = require('amqplib');
const { v4: uuidv4 } = require('uuid');

class AmqpComplianceTest {
    constructor(host = 'localhost', port = 5672) {
        this.host = host;
        this.port = port;
        this.results = [];
        this.connection = null;
        this.channel = null;
    }

    async connect() {
        const url = `amqp://guest:guest@${this.host}:${this.port}`;
        this.connection = await amqp.connect(url);
        this.channel = await this.connection.createChannel();
    }

    async disconnect() {
        if (this.channel) await this.channel.close().catch(() => {});
        if (this.connection) await this.connection.close().catch(() => {});
    }

    async runTest(name, testFunc) {
        try {
            await testFunc();
            this.results.push({ test: name, status: 'PASS', error: null });
            console.log(`  [PASS] ${name}`);
            return true;
        } catch (e) {
            this.results.push({ test: name, status: 'FAIL', error: e.message });
            console.log(`  [FAIL] ${name}: ${e.message}`);
            return false;
        }
    }

    // ==================== Connection Tests ====================

    async testConnectionOpen() {
        if (!this.connection) throw new Error('Connection should be open');
    }

    async testChannelOpen() {
        if (!this.channel) throw new Error('Channel should be open');
    }

    // ==================== Queue Tests ====================

    async testQueueDeclare() {
        const queueName = `test.nodejs.queue.${uuidv4().substring(0, 8)}`;
        const result = await this.channel.assertQueue(queueName, { autoDelete: true });
        if (result.queue !== queueName) throw new Error('Queue name mismatch');
    }

    async testQueueDeclareAnonymous() {
        const result = await this.channel.assertQueue('', { exclusive: true });
        if (!result.queue || result.queue === '') throw new Error('Anonymous queue should have a name');
    }

    // ==================== Exchange Tests ====================

    async testExchangeDeclareDirect() {
        const exchangeName = `test.nodejs.direct.${uuidv4().substring(0, 8)}`;
        await this.channel.assertExchange(exchangeName, 'direct', { autoDelete: true });
    }

    async testExchangeDeclareFanout() {
        const exchangeName = `test.nodejs.fanout.${uuidv4().substring(0, 8)}`;
        await this.channel.assertExchange(exchangeName, 'fanout', { autoDelete: true });
    }

    async testExchangeDeclareTopic() {
        const exchangeName = `test.nodejs.topic.${uuidv4().substring(0, 8)}`;
        await this.channel.assertExchange(exchangeName, 'topic', { autoDelete: true });
    }

    // ==================== Binding Tests ====================

    async testQueueBind() {
        const exchangeName = `test.nodejs.bind.ex.${uuidv4().substring(0, 8)}`;
        const queueName = `test.nodejs.bind.q.${uuidv4().substring(0, 8)}`;

        await this.channel.assertExchange(exchangeName, 'direct', { autoDelete: true });
        await this.channel.assertQueue(queueName, { autoDelete: true });
        await this.channel.bindQueue(queueName, exchangeName, 'test.key');
    }

    // ==================== Publish/Consume Tests ====================

    async testBasicPublishConsume() {
        const queueName = `test.nodejs.pubsub.${uuidv4().substring(0, 8)}`;
        const testMessage = `Hello from Node.js! ${uuidv4()}`;

        await this.channel.assertQueue(queueName, { autoDelete: true });
        this.channel.sendToQueue(queueName, Buffer.from(testMessage));

        const received = await new Promise((resolve, reject) => {
            const timeout = setTimeout(() => reject(new Error('Timeout waiting for message')), 5000);
            this.channel.consume(queueName, (msg) => {
                if (msg) {
                    clearTimeout(timeout);
                    this.channel.ack(msg);
                    resolve(msg.content.toString());
                }
            });
        });

        if (received !== testMessage) throw new Error(`Message mismatch: ${received} !== ${testMessage}`);
    }

    async testPublishWithProperties() {
        const queueName = `test.nodejs.props.${uuidv4().substring(0, 8)}`;

        await this.channel.assertQueue(queueName, { autoDelete: true });
        this.channel.sendToQueue(queueName, Buffer.from('test'), {
            contentType: 'text/plain',
            correlationId: 'corr-123',
            messageId: 'msg-456',
            priority: 5
        });

        const msg = await new Promise((resolve, reject) => {
            const timeout = setTimeout(() => reject(new Error('Timeout waiting for message')), 5000);
            this.channel.consume(queueName, (m) => {
                if (m) {
                    clearTimeout(timeout);
                    this.channel.ack(m);
                    resolve(m);
                }
            });
        });

        if (msg.properties.contentType !== 'text/plain') throw new Error('contentType mismatch');
        if (msg.properties.correlationId !== 'corr-123') throw new Error('correlationId mismatch');
        if (msg.properties.messageId !== 'msg-456') throw new Error('messageId mismatch');
    }

    async testPublishWithHeaders() {
        const queueName = `test.nodejs.headers.${uuidv4().substring(0, 8)}`;

        await this.channel.assertQueue(queueName, { autoDelete: true });
        this.channel.sendToQueue(queueName, Buffer.from('test'), {
            headers: {
                'x-custom-header': 'custom-value',
                'x-number': 42
            }
        });

        const msg = await new Promise((resolve, reject) => {
            const timeout = setTimeout(() => reject(new Error('Timeout waiting for message')), 5000);
            this.channel.consume(queueName, (m) => {
                if (m) {
                    clearTimeout(timeout);
                    this.channel.ack(m);
                    resolve(m);
                }
            });
        });

        if (!msg.properties.headers) throw new Error('Headers should exist');
        if (msg.properties.headers['x-custom-header'] !== 'custom-value') {
            throw new Error(`Header mismatch: ${msg.properties.headers['x-custom-header']}`);
        }
    }

    async testMessageOrdering() {
        const queueName = `test.nodejs.order.${uuidv4().substring(0, 8)}`;
        const messageCount = 10;

        await this.channel.assertQueue(queueName, { autoDelete: true });

        for (let i = 0; i < messageCount; i++) {
            this.channel.sendToQueue(queueName, Buffer.from(`Message-${i}`));
        }

        const received = [];
        await new Promise((resolve, reject) => {
            const timeout = setTimeout(() => reject(new Error('Timeout waiting for messages')), 10000);
            this.channel.consume(queueName, (msg) => {
                if (msg) {
                    received.push(msg.content.toString());
                    this.channel.ack(msg);
                    if (received.length === messageCount) {
                        clearTimeout(timeout);
                        resolve();
                    }
                }
            });
        });

        for (let i = 0; i < messageCount; i++) {
            if (received[i] !== `Message-${i}`) {
                throw new Error(`Out of order at ${i}: ${received[i]}`);
            }
        }
    }

    // ==================== Acknowledgment Tests ====================

    async testBasicNackRequeue() {
        const queueName = `test.nodejs.nack.${uuidv4().substring(0, 8)}`;
        let deliveryCount = 0;

        await this.channel.assertQueue(queueName, { autoDelete: true });
        this.channel.sendToQueue(queueName, Buffer.from('nack-me'));

        await new Promise((resolve, reject) => {
            const timeout = setTimeout(() => reject(new Error('Timeout')), 5000);
            this.channel.consume(queueName, (msg) => {
                if (msg) {
                    deliveryCount++;
                    if (deliveryCount === 1) {
                        this.channel.nack(msg, false, true); // requeue
                    } else {
                        this.channel.ack(msg);
                        clearTimeout(timeout);
                        resolve();
                    }
                }
            });
        });

        if (deliveryCount !== 2) throw new Error(`Expected 2 deliveries, got ${deliveryCount}`);
    }

    // ==================== QoS Tests ====================

    async testBasicQos() {
        await this.channel.prefetch(5);
        const queueName = `test.nodejs.qos.${uuidv4().substring(0, 8)}`;
        await this.channel.assertQueue(queueName, { autoDelete: true });

        for (let i = 0; i < 10; i++) {
            this.channel.sendToQueue(queueName, Buffer.from(`Message-${i}`));
        }
    }

    // ==================== Exchange Routing Tests ====================

    async testDirectRouting() {
        const exchangeName = `test.nodejs.routing.direct.${uuidv4().substring(0, 8)}`;
        const queue1 = `test.nodejs.routing.q1.${uuidv4().substring(0, 8)}`;
        const queue2 = `test.nodejs.routing.q2.${uuidv4().substring(0, 8)}`;

        await this.channel.assertExchange(exchangeName, 'direct', { autoDelete: true });
        await this.channel.assertQueue(queue1, { autoDelete: true });
        await this.channel.assertQueue(queue2, { autoDelete: true });
        await this.channel.bindQueue(queue1, exchangeName, 'key1');
        await this.channel.bindQueue(queue2, exchangeName, 'key2');

        this.channel.publish(exchangeName, 'key1', Buffer.from('to-q1'));
        this.channel.publish(exchangeName, 'key2', Buffer.from('to-q2'));

        const receivedQ1 = [];
        const receivedQ2 = [];

        await new Promise((resolve, reject) => {
            const timeout = setTimeout(() => reject(new Error('Timeout')), 5000);
            let count = 0;

            this.channel.consume(queue1, (msg) => {
                if (msg) {
                    receivedQ1.push(msg.content.toString());
                    this.channel.ack(msg);
                    if (++count === 2) { clearTimeout(timeout); resolve(); }
                }
            });

            this.channel.consume(queue2, (msg) => {
                if (msg) {
                    receivedQ2.push(msg.content.toString());
                    this.channel.ack(msg);
                    if (++count === 2) { clearTimeout(timeout); resolve(); }
                }
            });
        });

        if (receivedQ1[0] !== 'to-q1') throw new Error('Direct routing to q1 failed');
        if (receivedQ2[0] !== 'to-q2') throw new Error('Direct routing to q2 failed');
    }

    async testFanoutRouting() {
        const exchangeName = `test.nodejs.routing.fanout.${uuidv4().substring(0, 8)}`;
        const queue1 = `test.nodejs.fanout.q1.${uuidv4().substring(0, 8)}`;
        const queue2 = `test.nodejs.fanout.q2.${uuidv4().substring(0, 8)}`;

        await this.channel.assertExchange(exchangeName, 'fanout', { autoDelete: true });
        await this.channel.assertQueue(queue1, { autoDelete: true });
        await this.channel.assertQueue(queue2, { autoDelete: true });
        await this.channel.bindQueue(queue1, exchangeName, '');
        await this.channel.bindQueue(queue2, exchangeName, '');

        this.channel.publish(exchangeName, '', Buffer.from('broadcast'));

        const receivedQ1 = [];
        const receivedQ2 = [];

        await new Promise((resolve, reject) => {
            const timeout = setTimeout(() => reject(new Error('Timeout')), 5000);
            let count = 0;

            this.channel.consume(queue1, (msg) => {
                if (msg) {
                    receivedQ1.push(msg.content.toString());
                    this.channel.ack(msg);
                    if (++count === 2) { clearTimeout(timeout); resolve(); }
                }
            });

            this.channel.consume(queue2, (msg) => {
                if (msg) {
                    receivedQ2.push(msg.content.toString());
                    this.channel.ack(msg);
                    if (++count === 2) { clearTimeout(timeout); resolve(); }
                }
            });
        });

        if (receivedQ1[0] !== 'broadcast') throw new Error('Fanout to q1 failed');
        if (receivedQ2[0] !== 'broadcast') throw new Error('Fanout to q2 failed');
    }

    async runAllTests() {
        console.log('\n' + '='.repeat(60));
        console.log('Node.js amqplib AMQP 0-9-1 Compliance Tests');
        console.log(`Server: ${this.host}:${this.port}`);
        console.log('='.repeat(60) + '\n');

        const tests = [
            ['connection_open', () => this.testConnectionOpen()],
            ['channel_open', () => this.testChannelOpen()],
            ['queue_declare', () => this.testQueueDeclare()],
            ['queue_declare_anonymous', () => this.testQueueDeclareAnonymous()],
            ['exchange_declare_direct', () => this.testExchangeDeclareDirect()],
            ['exchange_declare_fanout', () => this.testExchangeDeclareFanout()],
            ['exchange_declare_topic', () => this.testExchangeDeclareTopic()],
            ['queue_bind', () => this.testQueueBind()],
            ['basic_publish_consume', () => this.testBasicPublishConsume()],
            ['publish_with_properties', () => this.testPublishWithProperties()],
            ['publish_with_headers', () => this.testPublishWithHeaders()],
            ['message_ordering', () => this.testMessageOrdering()],
            ['basic_nack_requeue', () => this.testBasicNackRequeue()],
            ['basic_qos', () => this.testBasicQos()],
            ['direct_routing', () => this.testDirectRouting()],
            ['fanout_routing', () => this.testFanoutRouting()],
        ];

        let passed = 0;
        let failed = 0;

        for (const [name, testFunc] of tests) {
            if (await this.runTest(name, testFunc)) {
                passed++;
            } else {
                failed++;
            }
        }

        console.log('\n' + '='.repeat(60));
        console.log(`Results: ${passed} passed, ${failed} failed`);
        console.log('='.repeat(60) + '\n');

        return { passed, failed, results: this.results };
    }
}

async function main() {
    const host = process.argv[2] || 'localhost';
    const port = parseInt(process.argv[3]) || 5672;

    const test = new AmqpComplianceTest(host, port);

    try {
        console.log(`Connecting to AMQP server at ${host}:${port}...`);
        await test.connect();
        const results = await test.runAllTests();

        console.log('\n--- JSON RESULTS ---');
        console.log(JSON.stringify(results));

        process.exit(results.failed === 0 ? 0 : 1);
    } catch (e) {
        console.error(`Connection failed: ${e.message}`);
        process.exit(1);
    } finally {
        await test.disconnect();
    }
}

main();
